src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
             return validatedUri.PartUriString;
         }
 
+        //This method validates the packUri and returns its two components if they are valid-
+        //1. Package Uri
+        //2. Part Uri
+        internal static void ValidateAndGetPackUriComponents(Uri packUri, out Uri packageUri, out Uri partUri)
+        {
+            //Validate if its not null and is an absolute Uri, has pack:// Scheme.
+            packUri = ValidatePackUri(packUri);
+            packageUri = GetPackageUriComponent(packUri);
+            partUri = GetPartUriComponent(packUri);
+        }
+
         #endregion Internal Methods
-        
+
+        #region Private Constructor
+
+        static PackUriHelper()
+        {
+            EnsurePackSchemeRegistered();
+        }
+
+        #endregion Private Constructor
+
         #region Private Methods
 
         private static readonly char[] HexUpperChars = {
             '0', '1', '2', '3', '4', '5', '6', '7',
             '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
         };
 
+        /// <summary>
+        /// This method is used to create a valid pack Uri
+        /// </summary>
+        /// <param name="packageUri">This is the uri that points to the entire package.
+        /// This parameter should be an absolute Uri. This parameter cannot be null or empty </param>
+        /// <param name="partUri">This is the uri that points to the part within the package
+        /// This parameter should be a relative Uri.
+        /// This parameter can be null in which case we will create a valid pack uri
+        /// that references the entire package</param>
+        /// <param name="fragment">Fragment for the resulting Pack URI. This parameter can be null
+        /// The fragment string must start with a "#"</param>
+        /// <returns>A Uri with the "pack://" scheme</returns>
+        /// <exception cref="ArgumentNullException">If packageUri parameter is null</exception>
+        /// <exception cref="ArgumentException">If packageUri parameter is not an absolute Uri</exception>
+        /// <exception cref="ArgumentException">If partUri parameter does not conform to the valid partUri syntax</exception>
+        /// <exception cref="ArgumentException">If fragment parameter is empty or does not start with a "#"</exception>
+        private static Uri Create(Uri packageUri, Uri partUri, string fragment)
```

Comment:
Maybe I'm missing it, but is this ever called with a non-null fragment? If no, it would seem we could just assert that and then delete the dead code below for the non-null case.

Reply:
I had it in here https://github.com/dotnet/corefx/issues/29959, but it wasn't hit when you ripped the code out of WindowsBase, so my guess is that it wasn't used by WindowsBase?  I'm OK adding it back as public, we shouldn't subset the public API unless we have a good reason.

Reply:
sounds good, i'll make it public and add a test for it